### PR TITLE
Use searchCaseReference as the target case reference when attaching an exception record

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1275

### Change description ###
- When attaching an Exception Record to an existing case, 
   - Validate `searchCaseReference` value in the callback request as the target case reference.
   - Validate  `attachToCaseReference` value when `searchCaseReference`  value doesn't exist (so 
   that the current functionality will not be broken)

- Added a functional test to verify if `searchCaseReference` takes priority when both  `searchCaseReference` and `attachToCaseReference` have values in the callback request.

**Next steps:**
- update Exception record definitions to include searchCaseReference field
- Update callback service to attach the Exception Record to `searchCaseReference` and 
remove validation on `attachToCaseReference`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java [hunk 1]
```diff
@@ -193,6 +193,39 @@ public void should_attach_exception_record_to_case_by_attach_to_case_reference()
         verifyExceptionRecordAttachesToCase(null);
     }
 
+    @Test
+    public void should_attach_exception_record_to_case_by_searchCaseReference_and_attachToCaseReference()
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:197]: how are both `searchCaseReference` and `attachToCaseReference` going to be set ? Is it by CCD but I am not sure how both fields will be populated

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java [hunk 2]
```diff
@@ -193,6 +193,39 @@ public void should_attach_exception_record_to_case_by_attach_to_case_reference()
...
+        throws Exception {
+        //given
+        // set searchCaseReference as targetCase id
+        CaseDetails targetCase = ccdCaseCreator.createCase(emptyList(), Instant.now());
+
+        // set attachToCaseReference value with attachToCaseDetails id, which should be ignored
+        CaseDetails attachToCaseDetails = ccdCaseCreator.createCase(emptyList(), Instant.now());
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:204]: not sure comments and code are in sync. Where are we setting the `searchCaseReference` value

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java (hunk 1 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:197]: yeah, it's not realistic. I'll update the test to send `searchCaseReference` value, `searchCaseReferenceType` as null and without `attachToCaseReference` field. Is that ok?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java (hunk 2 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:204]: I'll update the test

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java [hunk 3]
```diff
@@ -312,6 +340,37 @@ private Response invokeCallbackEndpoint(
             .post("/callback/attach_case?ignore-warning=true");
     }
 
+    private Response invokeCallbackEndpointWithSearchCaseRefAndAttachToCaseRef(
+        CaseDetails searchCaseRefCaseDetails,
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:344]: do you need to pass whole object here ? just case id is used so may be just pass that ?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java [hunk 4]
```diff
@@ -312,6 +340,37 @@ private Response invokeCallbackEndpoint(
...
+            .given()
+            .relaxedHTTPSValidation()
+            .baseUri(testUrl)
+            .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
+            .header(SyntheticHeaders.SYNTHETIC_TEST_SOURCE, "Bulk Scan Orchestrator Functional test")
+            .header(AUTHORIZATION, ccdAuthenticator.getUserToken())
+            .header(CcdCallbackController.USER_ID, ccdAuthenticator.getUserDetails().getId())
+            .body(callbackRequest)
+            .when()
+            .post("/callback/attach_case?ignore-warning=true");
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:371]: warning can be false here ? No warnings list is populated so ignore warning = true would not be the right scenario

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java [hunk 5]
```diff
@@ -193,6 +193,34 @@ public void should_attach_exception_record_to_case_by_attach_to_case_reference()
         verifyExceptionRecordAttachesToCase(null);
     }
 
+    @Test
+    public void should_attach_exception_record_to_case_by_search_case_reference()
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:197]: Also do we have a test case where both searchCaseReference and attachToCaseReference are set in call back ?

Comment (nitinprabhu) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/AttachExceptionRecordToExistingCaseTest.java:197]: Realised this is not needed

**Pull request merged**